### PR TITLE
Add load balancer

### PR DESCRIPTION
**Pull request by EugeneGal**
Andrei, Dima, plz review from the SOLID, concurrency issues perspectives. Probably you will advise to enhance this solution and make it more robust. Waiting for you feedback!

### CODE REVIEW

File: other/coding/src/main/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategy.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class RoundRobinServerPickStrategy implements ServerPickStrategy {
+
+    private final AtomicInteger counter = new AtomicInteger(0);
+
+    @Override
+    public Optional<Server> pick(List<Server> servers) {
+        List<Server> snapshot = new ArrayList<>(servers);
```
Comment (tinyzero4) [inline other/coding/src/main/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategy.java]: is that defensive copy? what if a client would like to use it's own picker strategy? how would you enforce him to make a copy?

File: other/coding/src/main/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategy.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        if (snapshot.isEmpty()) {
```
Comment (tinyzero4) [inline other/coding/src/main/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategy.java]: you can validate your input argument

File: other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    public boolean addServer(Server server) {
+        return servers.add(server);
+    }
+
+    public boolean removeServer(int id) {
+        return servers.removeIf(server -> server.getId() == id);
+    }
+
+    public void handleRequest(Request request) {
```
Comment (tinyzero4) [inline other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java]: I think there will be a sequence of small tasks/improvements, but the main goal will be to design a high quality, testable, extensible,supportable api e.g. for picking next server If you handle request i bet they will ask for a return type for such an op.
Think more like a client lib and ask Q

File: other/coding/src/main/java/org/example/coding/loadbalancer/model/Request.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package org.example.coding.loadbalancer.model;
+
+public class Request {
```
Comment (tinyzero4) [inline other/coding/src/main/java/org/example/coding/loadbalancer/model/Request.java]: I would also mark it as final

File: other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+
+public class LoadBalancerTest {
+
+    @Test
+    void shouldHandleRequest() {
+        Server server = spy(new Server(1, "url1"));
```
Comment (tinyzero4) [inline other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java]: why spy?

File: other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        List<Server> servers = List.of(server);
+        ServerPickStrategy pickStrategy = spy(new RoundRobinServerPickStrategy());
```
Comment (tinyzero4) [inline other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java]: Just use mock ServerPickStrategy

File: other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        ServerPickStrategy strategy = new RoundRobinServerPickStrategy();
+
+        Server server1 = new Server(1, "url1");
+        Server server2 = new Server(2, "url2");
+        Server server3 = new Server(3, "url3");
+
+        LoadBalancer loadBalancer = new LoadBalancer(new ArrayList<>(List.of(server1, server2, server3)), strategy);
+
+        CountDownLatch countDownLatch = new CountDownLatch(5000);
+        AtomicInteger failedThreadsCount = new AtomicInteger(0);
```
Comment (tinyzero4) [inline other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java:103]: I doubt you will have time for such a neat test, just polish your api and talk through concurrency

File: other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            thread.start();
+        }
+
+        try {
+            countDownLatch.await();
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+        // Remove server in the middle of execution to get possible IndexOutOfBoundsException in RoundRobinServerPickStrategy#pick()
+        loadBalancer.removeServer(3);
```
Comment (tinyzero4) [inline other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java]: possible, the many `possible` constructs you code the many questions you will get

File: other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+        for (Thread thread : threads) {
+            try {
+                thread.join();
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
```
Comment (tinyzero4) [inline other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java:131]: you will get a Q regarding this handling of Interrupted exception for sure, ask if you need to code concurrency tests beforehand.

File: other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+            FutureTask<Server> futureTask = new FutureTask<>(callable);
+            futureTasks.add(futureTask);
+            new Thread(futureTask).start();
+        }
+
+        int server1Count = 0, server2Count = 0, server3Count = 0;
+        for (FutureTask<Server> futureTask : futureTasks) {
+            try {
+                Server server = futureTask.get();
+                if (server.equals(server1)) {
```
Comment (tinyzero4) [inline other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java:37]: Btw, to equals/hashcode in Server.class, be prepared for questions

File: other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+            }
+        }
+        assertEquals(3333, server1Count);
+        assertEquals(3333, server2Count);
+        assertEquals(3333, server3Count);
+    }
+
+    private static class SelectServerCallable implements Callable<Server> {
+
+        private CyclicBarrier cyclicBarrier;
```
Comment (tinyzero4) [inline other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java:55]: overkill, ask if they want concurrency tests, you will be out of time for doing this

File: other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+            Request request = new Request(i);
+            Runnable runnable = () -> {
+                loadBalancer.handleRequest(request);
+                countDownLatch.countDown();
+            };
+
+            Thread thread = new Thread(runnable);
+            thread.setUncaughtExceptionHandler((t, e) -> failedThreadsCount.incrementAndGet());
+            threads.add(thread);
+            thread.start();
```
Comment (tinyzero4) [inline other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java:116]: even if you are forced to run a concurrency test don't use Thread, you will get a Q why not pool

File: other/coding/src/main/java/org/example/coding/loadbalancer/model/Server.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
+package org.example.coding.loadbalancer.model;
+
+public class Server {
+
+    private final int id;
+
+    private final String url;
+
+    public Server(int id, String url) {
+        this.id = id;
```
Comment (tinyzero4) [inline other/coding/src/main/java/org/example/coding/loadbalancer/model/Server.java]: no validation

File: other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+public class LoadBalancer {
+
+    private final List<Server> servers = new CopyOnWriteArrayList<>();
+
+    private final ServerPickStrategy pickStrategy;
+
+    public LoadBalancer(List<Server> servers, ServerPickStrategy pickStrategy) {
```
Comment (tinyzero4) [inline other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java]: no validation

File: other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+                } else if (server.equals(server2)) {
+                    server2Count++;
+                } else {
+                    server3Count++;
+                }
+            } catch (InterruptedException | ExecutionException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        assertEquals(3333, server1Count);
```
Comment (tinyzero4) [inline other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java:48]: I don't understand why 3333? The main purpose of RR is to load balance evenly one by one. There are no tests for functional requirements, e.g. first s1, then s2, then s3, then loopback. It coulld be that at first you will get 3333 s1, than 3333 s2, then 3333 s3, is that a RR?

File: other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+        public SelectServerCallable(CyclicBarrier cyclicBarrier, List<Server> servers, ServerPickStrategy strategy) {
+            this.cyclicBarrier = cyclicBarrier;
+            this.servers = servers;
+            this.strategy = strategy;
+        }
+
+        @Override
+        public Server call() {
+            try {
+                cyclicBarrier.await();
```
Comment (tinyzero4) [inline other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java:70]: Why do you need it at all?

**Review by tinyzero4 — CHANGES_REQUESTED** [review]
rejected, needs work :)

File: other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java (hunk 3 above)
Comment (EugeneGal) [reply other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java]: Do you mean that there should be a return value from this `handleRequest()` method instead of void? What could you suggest?

Comment (efan) [reply other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java]: If the task specifically mentions something about `get()` instead of handle request, then you'll know for sure you don't need `handleRequest`. If not, probably could be simplified to just having get and forget about request when discussing.

File: other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java (hunk 14 above)
Comment (efan) [reply other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java]: Overall too few validations everywhere. 

There should be plenty of corner cases, need to mention at least some of them, e.g. how many maximum servers are allowed to be in load balancer? What if servers list is null? What if it is empty? What if pickStrategy is null? 

Might be worth mentioning that in java there are assertions, but should typically be **not** on public methods, i.e. we could do 
```
assert pickStrategy != null
```
But yeah, nobody is using those :) (should be enabled on jvm i think?)

In this case i think simple 
```
java.util.Objects.requireNonNull(pickStrategy, "pickStrategy arg should not be null");
```
should be sufficient. However as it still throws null pointer, a better way could be to throw `IllegalArgumentException`

All this exceptions will then need to be tested with junit tests.

File: other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java (hunk 11 above)
Comment (efan) [reply other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java:55]: I think it looks quite complicated, so probably won't be worth writing it right away, might kill a lot of time. A friend said they asked him about concurrent tests. In this case it's good to be able to discuss this and even write some code if needed. Also need to be familiar with different classes in `java.util.concurrent` package. But cyclic barrier is probably not the first choice :)

It probably will be more important to get the public api right and implementation with regards to that api as @tinyzero4 already mentioned. For example already mentioned `pickStrategy = null` just blows up all implementation, i.e. it's not checked when load balancer class is created and then when actual `pickStrategy.pick` is called it just fails with null pointer. In practice this means the error won't be caught in build time(as there's no tests), it also won't be caught at app startup time(no check when load balancer is created), and only will appear when app is already deployed.

In short, my opinion, don't go into too complicated stuff if not asked, get simpler things right. Maybe just mention that this is possible, be ready to discuss and write something if asked to write.

File: other/coding/src/main/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategy.java [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class RoundRobinServerPickStrategy implements ServerPickStrategy {
+
+    private final AtomicInteger counter = new AtomicInteger(0);
+
+    @Override
+    public Optional<Server> pick(List<Server> servers) {
```
Comment (efan) [inline other/coding/src/main/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategy.java:14]: Not sure if this is valid from interview point of view, but for me personally it would be curious to also discuss the choices with regards to SOLID. 

For example, do we really need servers there? Basically it just gives us a number from 0 to servers.size(). Then whole story of sending list and copying etc. does it even make sense? Could be pick server index for example.

But if we need servers, could those also be a field in the class? i.e. design this more like RoundRobinServerList of some kind with .add, .remove, .nextServer. If so then do we really need a separate strategy class, i.e. what if we just have AbstractLoadBalancer and then this RoundRobinServerList just becomes RoundRobinLoadBalancer specific implementation without any strategy.

In other words inheritance vs composition. Like usually the rule is composition over inheritance, i.e. strategy would be preferred, but when we go for inheritance for example?

In this case what's the responsibility of strategy vs responsibility of load balancer? Seems like load balancer does handle request, strategy does picking of server. In case if they go with the task where there's no handle request, just get method. Does it change anything and what will be responsibility of load balancer in that case.

File: other/coding/src/main/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategy.java (hunk 1 above)
Comment (EugeneGal) [reply other/coding/src/main/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategy.java]: removed

File: other/coding/src/main/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategy.java (hunk 2 above)
Comment (EugeneGal) [reply other/coding/src/main/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategy.java]: done

File: other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java (hunk 3 above)
Comment (EugeneGal) [reply other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java]: done

File: other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java (hunk 5 above)
Comment (EugeneGal) [reply other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java]: Now no need in it, removed

File: other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java (hunk 6 above)
Comment (EugeneGal) [reply other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java]: removed

File: other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java (hunk 9 above)
Comment (EugeneGal) [reply other/coding/src/test/java/org/example/coding/loadbalancer/LoadBalancerTest.java:131]: Ok, thanks!

File: other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java (hunk 10 above)
Comment (EugeneGal) [reply other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java:37]: Ok, thanks!

File: other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java (hunk 11 above)
Comment (EugeneGal) [reply other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java:55]: Guys, thanks for advise!

File: other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java (hunk 14 above)
Comment (EugeneGal) [reply other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java]: Got it, added more validation at corner cases, thanks!

File: other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java (hunk 16 above)
Comment (EugeneGal) [reply other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java:70]: To test that `AtomicInteger counter = new AtomicInteger(0)` inside `RoundRobinServerPickStrategy` in concurrent env works as expected. So, for example, if counter was an int type, due to race conditions it would work inappropriately.

File: other/coding/src/main/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategy.java (hunk 17 above)
Comment (EugeneGal) [reply other/coding/src/main/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategy.java:14]: For now list copy was removed. I pass list of servers since there are might be other implementations of `ServerPickStrategy` and they might require more information (not only count of servers like in `RoundRobinServerPickStrategy`) to choose appropriate server. So having full server object in contract will make it easier for adding other more complicated implementations.

File: other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+        }
+        this.pickStrategy = pickStrategy;
+    }
+
+    public boolean registerServer(Server server) {
+        if (server == null) {
+            return false;
+        }
+
+        synchronized (this) {
```
Comment (efan) [inline other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java:31]: why synchronised if map is concurrent?

File: other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+            if (urlToServerMap.containsKey(server.getUrl())) {
+                return false;
+            }
+
+            if (urlToServerMap.size() >= MAX_CAPACITY) {
+                return false;
+            }
+
+            return urlToServerMap.put(server.getUrl(), server) == null;
```
Comment (efan) [inline other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java]: Concurrent map should have something like `putIfAbsent`

File: other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+
+        return pickStrategy.pick(new ArrayList<>(urlToServerMap.values()));
+    }
+
+    public boolean removeServer(Server server) {
+        if (server == null || server.getUrl() == null) {
+            return false;
+        }
+
+        return server.equals(urlToServerMap.remove(server.getUrl()));
```
Comment (efan) [inline other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java]: looks complicated too me, do we even need `boolean` as return value for add/remove? like null checks could throw illegal argument exception and then if no exception add/remove is successful

File: other/coding/src/main/java/org/example/coding/loadbalancer/model/Server.java [hunk 21]
```diff
@@ -0,0 +1,33 @@
+package org.example.coding.loadbalancer.model;
+
+import java.util.Objects;
+
+public class Server {
+
+    private final String url;
+
+    public Server(String url) {
+        this.url = url;
```
Comment (efan) [inline other/coding/src/main/java/org/example/coding/loadbalancer/model/Server.java:10]: Do we need any validation for url? Is string good enough, or maybe we should go for java.net.URL?

File: other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java (hunk 18 above)
Comment (EugeneGal) [reply other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java:31]: There are several checks in sequence so there might be race conditions - so I think better to synchronise this piece of code and for adding server give access only to one thread

File: other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java (hunk 20 above)
Comment (EugeneGal) [reply other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java]: Changed to `return urlToServerMap.remove(server.getUrl(), server);`

File: other/coding/src/main/java/org/example/coding/loadbalancer/model/Server.java (hunk 21 above)
Comment (EugeneGal) [reply other/coding/src/main/java/org/example/coding/loadbalancer/model/Server.java:10]: For simplicity I used just String, but for sure can change to URL

File: other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java (hunk 19 above)
Comment (EugeneGal) [reply other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java]: done

File: other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java (hunk 18 above)
Comment (efan) [reply other/coding/src/main/java/org/example/coding/loadbalancer/LoadBalancer.java:31]: Right, valid point, i think i'd try to use something like `putIfAbsent` to avoid doing separate contains check. But not sure if something similar would be possible with size. My point was it seems concurrent functions of concurrent hash map are not used here at all. Then also there's no sync in remove method, which i'm not sure might or might not be a problem. Overall looks suspicious :)

File: other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java (hunk 15 above)
Comment (EugeneGal) [reply other/coding/src/test/java/org/example/coding/loadbalancer/pickstrategy/RoundRobinServerPickStrategyTest.java:48]: In round robin each server should be invoked the same number of times as its peers, otherwise it will be not round robin, I assume :)
